proof/invariant-abstract/ARM_HYP/ArchArch_AI.thy

```diff
@@ -983,24 +987,196 @@ lemma valid_global_vspace_mappings_vcpu_update_str:
+
+lemma associate_vcpu_tcb_valid_idle[wp]:
+  "\<lbrace>valid_idle and (\<lambda>s. tcb \<noteq> idle_thread s)\<rbrace> associate_vcpu_tcb vcpu tcb \<lbrace>\<lambda>_. valid_idle\<rbrace>"
+  apply (clarsimp simp: associate_vcpu_tcb_def)
+  by (wp hoare_vcg_all_lift
+      | wp (once) hoare_drop_imps
+      | wpc
+      | simp add: dissociate_vcpu_tcb_def vcpu_invalidate_active_def)+
+
+lemma arm_current_vcpu_update_valid_global_refs[simp]:
+  "valid_global_refs (s\<lparr>arch_state := arm_current_vcpu_update f (arch_state s)\<rparr>) = valid_global_refs s"
+  by (clarsimp simp: valid_global_refs_def global_refs_def
+              split: kernel_object.splits option.splits)
+
+crunches associate_vcpu_tcb
+  for valid_global_refs[wp]: valid_global_refs
+  (wp: crunch_wps)
+
+crunches vcpu_restore_reg
+  for valid_irq_states[wp]: valid_irq_states
+  (wp: crunch_wps dmo_valid_irq_states simp: writeVCPUHardwareReg_def)
+
+method repeat_unless methods stop cont
+  = (stop | (cont, repeat_unless stop cont))
+
+lemma is_irq_active_sp:
+  "\<lbrace>P\<rbrace> get_irq_state irq \<lbrace>\<lambda>rv s. P s \<and> (rv = interrupt_states s irq)\<rbrace>"
+  by (wpsimp simp: get_irq_state_def)
+
+\<comment> \<open>For forward reasoning in oare proofs, these lemmas allow skipping over the
+    left-hand-side of monadic bind, while keeping the same precondition.\<close>
+lemmas hoare_seq_ext_skip
+  = hoare_seq_ext[where B="\<lambda>_. A" and A=A for A, rotated]
+
+lemma restore_virt_timer_valid_irq_states[wp]:
+  "restore_virt_timer vcpu_ptr \<lbrace>valid_irq_states\<rbrace>"
+  apply (clarsimp simp: restore_virt_timer_def is_irq_active_def liftM_def)
+  apply (repeat_unless \<open>rule hoare_seq_ext[OF _ is_irq_active_sp]\<close>
+                       \<open>rule hoare_seq_ext_skip,
+                        wpsimp wp: dmo_valid_irq_states
+                             simp: isb_def setHCR_def set_cntv_cval_64_def read_cntpct_def
+                                   set_cntv_off_64_def\<close>)
+  apply (wpsimp simp: do_machine_op_def is_irq_active_def get_irq_state_def)
+  apply (clarsimp simp: valid_irq_states_def valid_irq_masks_def maskInterrupt_def in_monad)
+  done
+
+crunches vcpu_restore_reg_range, vcpu_enable
+  for valid_irq_states[wp]: valid_irq_states
+  (wp: crunch_wps dmo_valid_irq_states set_gic_vcpu_ctrl_hcr_irq_masks
+   simp: maskInterrupt_def isb_def setHCR_def)
+
+lemma vcpu_restore_valid_irq_states[wp]:
+  "vcpu_restore vcpu \<lbrace>valid_irq_states\<rbrace>"
+  unfolding vcpu_restore_def isb_def
+  by (wpsimp wp: dmo_valid_irq_states set_gic_vcpu_ctrl_hcr_irq_masks set_gic_vcpu_ctrl_vmcr_irq_masks
+                 set_gic_vcpu_ctrl_lr_irq_masks mapM_wp_inv set_gic_vcpu_ctrl_apr_irq_masks)
+
+crunches vcpu_switch
+  for valid_irq_states[wp]: valid_irq_states
+  (wp: crunch_wps dmo_valid_irq_states
+   simp: get_gic_vcpu_ctrl_lr_def get_gic_vcpu_ctrl_apr_def
+         get_gic_vcpu_ctrl_vmcr_def get_gic_vcpu_ctrl_hcr_def isb_def)
```

Comment:
Trying to understand crunch limitations here... wouldn't this be able to handle vcpu_restore_valid_irq_states as well, given the extra rules above? I think vcpu_switch calls vcpu_restore? I keep getting the feeling that some of these crunches could be compressed fairly easily. If this is copied from elsewhere, might not be worth the bother.